x-pack/filebeat/input/http_endpoint/handler_test.go

```diff
@@ -137,3 +144,109 @@ func Test_httpReadJSON(t *testing.T) {
 		})
 	}
 }
+
+type publisher struct {
+	events []beat.Event
+}
+
+func (p *publisher) Publish(event beat.Event) {
+	p.events = append(p.events, event)
+}
+
+func Test_apiResponse(t *testing.T) {
+	testCases := []struct {
+		name    string          // Sub-test name.
+		request *http.Request   // Input request.
+		events  []common.MapStr // Expected output events.
+	}{
+		{
+			name: "single event",
+			request: func() *http.Request {
+				req := httptest.NewRequest(http.MethodPost, "/", bytes.NewBufferString(`{"id":0}`))
+				req.Header.Set("Content-Type", "application/json")
+				return req
+			}(),
+			events: []common.MapStr{
+				{
+					"json": common.MapStr{
+						"id": int64(0),
+					},
+				},
+			},
+		},
+		{
+			name: "single event gzip",
+			request: func() *http.Request {
+				buf := new(bytes.Buffer)
+				b := gzip.NewWriter(buf)
+				_, _ = io.WriteString(b, `{"id":0}`)
+				b.Close()
+
+				req := httptest.NewRequest(http.MethodPost, "/", buf)
+				req.Header.Set("Content-Type", "application/json")
+				req.Header.Set("Content-Encoding", "gzip")
+				return req
+			}(),
+			events: []common.MapStr{
+				{
+					"json": common.MapStr{
+						"id": int64(0),
+					},
+				},
+			},
+		},
+		{
+			name: "multiple events gzip",
+			request: func() *http.Request {
+				events := []string{
+					`{"id":0}`,
+					`{"id":1}`,
+				}
+
+				buf := new(bytes.Buffer)
+				b := gzip.NewWriter(buf)
+				_, _ = io.WriteString(b, strings.Join(events, "\n"))
+				b.Close()
+
+				req := httptest.NewRequest(http.MethodPost, "/", buf)
+				req.Header.Set("Content-Type", "application/json")
+				req.Header.Set("Content-Encoding", "gzip")
+				return req
+			}(),
+			events: []common.MapStr{
+				{
+					"json": common.MapStr{
+						"id": int64(0),
+					},
+				},
+				{
+					"json": common.MapStr{
+						"id": int64(1),
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Setup
+			pub := new(publisher)
+			conf := defaultConfig()
+			apiHandler := newHandler(conf, pub, logp.NewLogger("http_endpoint.test"))
```

Comment:
This stanza could be put in the outer scope, but this makes sure the tests are isolated.

Reply:
Yeah, I figured isolation would be better. This way I don't have to reset the `publisher` state each time. I guess I could even mark the sub-tests with `t.Parallel()`.